### PR TITLE
Including throughputCount parameter to the extension.

### PR DESCRIPTION
**Pull request by SaranganJ**
## Purpose

If we initialize the performance extension at multiple places in the Siddhi code (In normal deployment mode) then each performance call will increment the 'eventcount' variable . That means for a single event ,the eventcount will be incremented by the number of times , which the performance extension is called.
Refer issue #853

## Approach

Include an additional parameter to the extension where the throughput count should be specified.So in case of multiple throughput calls only at once the 'eventCount' will be incremented.

### CODE REVIEW

File: modules/performance/extension/siddhi-execution-performance/component/src/main/java/org/wso2/extension/siddhi/execution/throughput/CalculatePerformanceStreamProcessorExtension.java [hunk 1]
```diff
@@ -77,7 +77,7 @@
                                 "from inputStream\n" +
                                 "select iijtimestamp,value\n" +
                                 "insert into tempStream;" +
-                                "from tempStream#throughput:throughput(iijtimestamp,value)\n" +
+                                "from tempStream#throughput:throughput(iijtimestamp,\"latency\",1)\n" +
```
Comment (Niveathika92) [inline modules/performance/extension/siddhi-execution-performance/component/src/main/java/org/wso2/extension/siddhi/execution/throughput/CalculatePerformanceStreamProcessorExtension.java]: Lets add @Parameter description for all parameters introduced

File: modules/performance/extension/siddhi-execution-performance/component/src/main/java/org/wso2/extension/siddhi/execution/throughput/CalculatePerformanceStreamProcessorExtension.java [hunk 2]
```diff
@@ -249,18 +255,25 @@ private static int getLogFileSequenceNumber() {
                         + "found " + attributeExpressionExecutors[1].getReturnType());
             }
 
-            if (attributeExpressionLength == 3) {
-                if (attributeExpressionExecutors[2].getReturnType() == Attribute.Type.INT) {
+            if (attributeExpressionExecutors[2].getReturnType() == Attribute.Type.INT) {
```
Comment (Niveathika92) [inline modules/performance/extension/siddhi-execution-performance/component/src/main/java/org/wso2/extension/siddhi/execution/throughput/CalculatePerformanceStreamProcessorExtension.java]: Why is the if clause empty?

File: modules/performance/extension/siddhi-execution-performance/component/src/main/java/org/wso2/extension/siddhi/execution/throughput/CalculatePerformanceStreamProcessorExtension.java [hunk 3]
```diff
@@ -249,18 +255,25 @@ private static int getLogFileSequenceNumber() {
...
-                            + "found " + attributeExpressionExecutors[1].getReturnType());
+                    throw new SiddhiAppValidationException("Fourth parameter expected to be int but "
+                            + "found " + attributeExpressionExecutors[3].getReturnType());
                 }
             }
 
         } else {
             throw new SiddhiAppValidationException("Input parameters for Log can be iijTimeStamp (Long), " +
-                    "type (String), recordwindow (int) but there are " +
+                    "type (String), throughputCount (int),  recordwindow (int) but there are " +
```
Comment (Niveathika92) [inline modules/performance/extension/siddhi-execution-performance/component/src/main/java/org/wso2/extension/siddhi/execution/throughput/CalculatePerformanceStreamProcessorExtension.java]: Is is throughput count or the order of the instance called?

File: modules/performance/extension/siddhi-execution-performance/component/src/main/java/org/wso2/extension/siddhi/execution/throughput/CalculatePerformanceStreamProcessorExtension.java [hunk 4]
```diff
@@ -306,10 +319,17 @@ private void calculateLatency(ComplexEventChunk<StreamEvent> streamEventChunk) {
...
                     long currentTime = System.currentTimeMillis();
-                    long iijTimestamp = (Long) (attributeExpressionExecutors[0].execute(streamEvent));
+
                     timeSpent += (currentTime - iijTimestamp);
-                    eventCount++;
-                    eventCountTotal++;
+
+                    long iijTimestamp = (Long) (attributeExpressionExecutors[0].execute(streamEvent));
+
+
```
Comment (Niveathika92) [inline modules/performance/extension/siddhi-execution-performance/component/src/main/java/org/wso2/extension/siddhi/execution/throughput/CalculatePerformanceStreamProcessorExtension.java]: Remove unnecessary new lines. Check in all places

File: modules/performance/extension/siddhi-execution-performance/component/src/main/java/org/wso2/extension/siddhi/execution/throughput/CalculatePerformanceStreamProcessorExtension.java [hunk 5]
```diff
@@ -91,10 +91,14 @@
...
                                 + "select iijtimestamp,value\n"
                                 + "insert into tempStream;\n"
-                                + "from tempStream#throughput:throughput(iijtimestamp,value,\"both\")\n"
+                                + "from tempStream#throughput:throughput(iijtimestamp,\"throughput\",2,120)\n"
                                 + "select \"aa\" as tempTrrb\n"
                                 + "insert into tempStream1;",
-                        description = "This is a filter query"
+                        description = "This is a filter query(Here third argument 2 indicates that " +
+                                "this throughput extension" +
+                                "is called for the second time in the query and fourth argement is an " +
```
Comment (miyurud) [inline modules/performance/extension/siddhi-execution-performance/component/src/main/java/org/wso2/extension/siddhi/execution/throughput/CalculatePerformanceStreamProcessorExtension.java]: Change "this throughput extension" to "this throughput extension "

File: modules/performance/extension/siddhi-execution-performance/component/src/main/java/org/wso2/extension/siddhi/execution/throughput/CalculatePerformanceStreamProcessorExtension.java [hunk 6]
```diff
@@ -91,10 +100,14 @@
...
                                 + "select iijtimestamp,value\n"
                                 + "insert into tempStream;\n"
-                                + "from tempStream#throughput:throughput(iijtimestamp,value,\"both\")\n"
+                                + "from tempStream#throughput:throughput(iijtimestamp,\"throughput\",2,120)\n"
                                 + "select \"aa\" as tempTrrb\n"
                                 + "insert into tempStream1;",
-                        description = "This is a filter query"
+                        description = "This is a filter query(Here third argument 2 indicates that " +
+                                "this throughput extension " +
+                                "is called for the second time in the query and fourth argement is an " +
```
Comment (miyurud) [inline modules/performance/extension/siddhi-execution-performance/component/src/main/java/org/wso2/extension/siddhi/execution/throughput/CalculatePerformanceStreamProcessorExtension.java]: Change "argement" to "argument"

File: modules/performance/extension/siddhi-execution-performance/component/src/main/java/org/wso2/extension/siddhi/execution/throughput/CalculatePerformanceStreamProcessorExtension.java [hunk 7]
```diff
@@ -91,10 +100,14 @@
...
+                                "optinal parameter of windowsize)" +
```
Comment (miyurud) [inline modules/performance/extension/siddhi-execution-performance/component/src/main/java/org/wso2/extension/siddhi/execution/throughput/CalculatePerformanceStreamProcessorExtension.java]: change optinal to optional

File: modules/performance/extension/siddhi-execution-performance/component/src/main/java/org/wso2/extension/siddhi/execution/throughput/CalculatePerformanceStreamProcessorExtension.java [hunk 8]
```diff
@@ -68,6 +68,15 @@
                 @Parameter(name = "iijtimestamp",
                            description = "This value used to find the sending timestamp from client",
                            type = {DataType.LONG}),
+
+                           name = "Count",
+                           description = "This value used to track the count of throughput extension calling",
```
Comment (miyurud) [inline modules/performance/extension/siddhi-execution-performance/component/src/main/java/org/wso2/extension/siddhi/execution/throughput/CalculatePerformanceStreamProcessorExtension.java]: This sentence "This value used to track the count of throughput extension calling" does not have meaning. What do you mean by count of throughput extension calling?

File: modules/performance/extension/siddhi-execution-performance/component/src/main/java/org/wso2/extension/siddhi/execution/throughput/CalculatePerformanceStreamProcessorExtension.java [hunk 9]
```diff
@@ -249,18 +262,23 @@ private static int getLogFileSequenceNumber() {
                         + "found " + attributeExpressionExecutors[1].getReturnType());
             }
 
-            if (attributeExpressionLength == 3) {
-                if (attributeExpressionExecutors[2].getReturnType() == Attribute.Type.INT) {
+            if (!(attributeExpressionExecutors[2].getReturnType() == Attribute.Type.INT)) {
```
Comment (miyurud) [inline modules/performance/extension/siddhi-execution-performance/component/src/main/java/org/wso2/extension/siddhi/execution/throughput/CalculatePerformanceStreamProcessorExtension.java]: Shouldn't this attribute be checked for whether it is a variable or constant? I see you have done that checking for the first two attributes but not for this attribute.

File: modules/performance/extension/siddhi-execution-performance/component/src/main/java/org/wso2/extension/siddhi/execution/throughput/CalculatePerformanceStreamProcessorExtension.java (hunk 3 above)
Comment (miyurud) [reply modules/performance/extension/siddhi-execution-performance/component/src/main/java/org/wso2/extension/siddhi/execution/throughput/CalculatePerformanceStreamProcessorExtension.java]: How do you get the values for executionType variable?

Comment (SaranganJ) [reply modules/performance/extension/siddhi-execution-performance/component/src/main/java/org/wso2/extension/siddhi/execution/throughput/CalculatePerformanceStreamProcessorExtension.java]: Type determines whether it is "throughput" or "Latency" or "Both" .It is the default parameter which originally in  the extension.i didn't change that.

File: modules/performance/extension/siddhi-execution-performance/component/src/main/java/org/wso2/extension/siddhi/execution/throughput/CalculatePerformanceStreamProcessorExtension.java (hunk 2 above)
Comment (SaranganJ) [reply modules/performance/extension/siddhi-execution-performance/component/src/main/java/org/wso2/extension/siddhi/execution/throughput/CalculatePerformanceStreamProcessorExtension.java]: Modified

**Comment by Niveathika92** [comment]
Pending till the code review:
Invitation: Code Review Meeting on SP Performance Extension @ Thu Nov 15, 2018 2pm - 2:30pm (IST)

File: modules/performance/extension/siddhi-execution-performance/component/src/main/java/org/wso2/extension/siddhi/execution/throughput/CalculatePerformanceStreamProcessorExtension.java (hunk 1 above)
Comment (SaranganJ) [reply modules/performance/extension/siddhi-execution-performance/component/src/main/java/org/wso2/extension/siddhi/execution/throughput/CalculatePerformanceStreamProcessorExtension.java]: Done

File: modules/performance/extension/siddhi-execution-performance/component/src/main/java/org/wso2/extension/siddhi/execution/throughput/CalculatePerformanceStreamProcessorExtension.java (hunk 4 above)
Comment (SaranganJ) [reply modules/performance/extension/siddhi-execution-performance/component/src/main/java/org/wso2/extension/siddhi/execution/throughput/CalculatePerformanceStreamProcessorExtension.java]: Acknowledged

File: modules/performance/extension/siddhi-execution-performance/component/src/main/java/org/wso2/extension/siddhi/execution/throughput/CalculatePerformanceStreamProcessorExtension.java (hunk 6 above)
Comment (SaranganJ) [reply modules/performance/extension/siddhi-execution-performance/component/src/main/java/org/wso2/extension/siddhi/execution/throughput/CalculatePerformanceStreamProcessorExtension.java]: Changed

File: modules/performance/extension/siddhi-execution-performance/component/src/main/java/org/wso2/extension/siddhi/execution/throughput/CalculatePerformanceStreamProcessorExtension.java (hunk 7 above)
Comment (SaranganJ) [reply modules/performance/extension/siddhi-execution-performance/component/src/main/java/org/wso2/extension/siddhi/execution/throughput/CalculatePerformanceStreamProcessorExtension.java]: Changed

File: modules/performance/extension/siddhi-execution-performance/component/src/main/java/org/wso2/extension/siddhi/execution/throughput/CalculatePerformanceStreamProcessorExtension.java (hunk 9 above)
Comment (SaranganJ) [reply modules/performance/extension/siddhi-execution-performance/component/src/main/java/org/wso2/extension/siddhi/execution/throughput/CalculatePerformanceStreamProcessorExtension.java]: Modified

File: modules/performance/extension/siddhi-execution-performance/component/src/main/java/org/wso2/extension/siddhi/execution/throughput/CalculatePerformanceStreamProcessorExtension.java [hunk 10]
```diff
@@ -249,18 +262,23 @@ private static int getLogFileSequenceNumber() {
...
+                throw new SiddhiAppValidationException("Third parameter expected to be int but "
+                        + "found " + attributeExpressionExecutors[2].getReturnType());
+            }
+
+            if (attributeExpressionLength == 4) {
+                if (attributeExpressionExecutors[3].getReturnType() == Attribute.Type.INT) {
```
[the remark this answers is not in the document]
Comment (SaranganJ) [reply modules/performance/extension/siddhi-execution-performance/component/src/main/java/org/wso2/extension/siddhi/execution/throughput/CalculatePerformanceStreamProcessorExtension.java]: Yes we should check

File: modules/performance/extension/siddhi-execution-performance/component/src/main/java/org/wso2/extension/siddhi/execution/throughput/CalculatePerformanceStreamProcessorExtension.java [hunk 11]
```diff
@@ -58,16 +58,32 @@
...
+
+                @Parameter(name = "type",
+                        description = "This value used to specify which metric value " +
+                                "you are going to exract(\"throughput\" " +
+                                "or \"latency\" or \"both\" )",
+                        type = {DataType.STRING}),
+
+                @Parameter(name = "windowsize",
+                        description = "This value used to determine for how " +
+                                "many events we are going to track the Metrics",
```
Comment (Niveathika92) [inline modules/performance/extension/siddhi-execution-performance/component/src/main/java/org/wso2/extension/siddhi/execution/throughput/CalculatePerformanceStreamProcessorExtension.java]: For optional parameter put optional = true

File: modules/performance/extension/siddhi-execution-performance/component/src/main/java/org/wso2/extension/siddhi/execution/throughput/CalculatePerformanceStreamProcessorExtension.java [hunk 12]
```diff
@@ -142,6 +161,8 @@ private static int setCompletedFlag(int sequenceNumber) {
         return 0;
     }
 
+
```
Comment (Niveathika92) [inline modules/performance/extension/siddhi-execution-performance/component/src/main/java/org/wso2/extension/siddhi/execution/throughput/CalculatePerformanceStreamProcessorExtension.java]: Remove unnecessary new lines. Check in all places

File: modules/performance/extension/siddhi-execution-performance/component/src/main/java/org/wso2/extension/siddhi/execution/throughput/CalculatePerformanceStreamProcessorExtension.java [hunk 13]
```diff
@@ -58,16 +58,32 @@
...
         description = "Measuring performance of stream processor with simple passthrough",
         parameters = {
                 @Parameter(name = "iijtimestamp",
-                           description = "This value used to find the sending timestamp from client",
-                           type = {DataType.LONG}),
+                        description = "This value used to find the sending timestamp from client",
+                        type = {DataType.LONG}),
+
+                @Parameter(name = "type",
+                        description = "This value used to specify which metric value " +
```
Comment (Niveathika92) [inline modules/performance/extension/siddhi-execution-performance/component/src/main/java/org/wso2/extension/siddhi/execution/throughput/CalculatePerformanceStreamProcessorExtension.java]: Is type also optional? If so what is the default value?

File: modules/performance/extension/siddhi-execution-performance/component/src/main/java/org/wso2/extension/siddhi/execution/throughput/CalculatePerformanceStreamProcessorExtension.java [hunk 14]
```diff
@@ -223,22 +244,24 @@ private static int getLogFileSequenceNumber() {
 
         siddhiAppContextName = siddhiAppContext.getName();
 
-        if (attributeExpressionLength == 2) {
+
+        if (attributeExpressionLength == 2 || attributeExpressionLength == 3 ||
+                attributeExpressionLength == 4) {
+
```
Comment (Niveathika92) [inline modules/performance/extension/siddhi-execution-performance/component/src/main/java/org/wso2/extension/siddhi/execution/throughput/CalculatePerformanceStreamProcessorExtension.java]: Use a loop with switch case for each attribute validation, for higher readabilty

File: modules/performance/extension/siddhi-execution-performance/component/src/main/java/org/wso2/extension/siddhi/execution/throughput/CalculatePerformanceStreamProcessorExtension.java [hunk 15]
```diff
@@ -535,17 +590,15 @@ private void createFile() {
...
-                                                                                    siddhiAppContextName + "-" +
-                                                                                    sequenceNumber + "-" +
-
-                                                                                   (outputFileTimeStamp)
-                                                                                   + ".csv")
-                                                                        .getAbsoluteFile()), StandardCharsets
-                                                     .UTF_8);
+                                                                            siddhiAppContextName + "-" +
+                                                                            sequenceNumber + "-" +
+                                                                            ".csv")
```
Comment (Niveathika92) [inline modules/performance/extension/siddhi-execution-performance/component/src/main/java/org/wso2/extension/siddhi/execution/throughput/CalculatePerformanceStreamProcessorExtension.java]: Why the id not added?

File: modules/performance/extension/siddhi-execution-performance/component/src/main/java/org/wso2/extension/siddhi/execution/throughput/CalculatePerformanceStreamProcessorExtension.java (hunk 11 above)
Comment (SaranganJ) [reply modules/performance/extension/siddhi-execution-performance/component/src/main/java/org/wso2/extension/siddhi/execution/throughput/CalculatePerformanceStreamProcessorExtension.java]: https://github.com/wso2/product-sp/pull/855/commits/3971112ad2ea08bef38437ea24bb12cad020692c

File: modules/performance/extension/siddhi-execution-performance/component/src/main/java/org/wso2/extension/siddhi/execution/throughput/CalculatePerformanceStreamProcessorExtension.java (hunk 15 above)
Comment (SaranganJ) [reply modules/performance/extension/siddhi-execution-performance/component/src/main/java/org/wso2/extension/siddhi/execution/throughput/CalculatePerformanceStreamProcessorExtension.java]: https://github.com/wso2/product-sp/pull/855/commits/3971112ad2ea08bef38437ea24bb12cad020692c

File: modules/performance/extension/siddhi-execution-performance/component/src/main/java/org/wso2/extension/siddhi/execution/throughput/CalculatePerformanceStreamProcessorExtension.java (hunk 12 above)
Comment (SaranganJ) [reply modules/performance/extension/siddhi-execution-performance/component/src/main/java/org/wso2/extension/siddhi/execution/throughput/CalculatePerformanceStreamProcessorExtension.java]: https://github.com/wso2/product-sp/pull/855/commits/3248156c1c90f9d63629b154810672b57e2b3fb9

File: modules/performance/extension/siddhi-execution-performance/component/src/main/java/org/wso2/extension/siddhi/execution/throughput/CalculatePerformanceStreamProcessorExtension.java (hunk 14 above)
Comment (SaranganJ) [reply modules/performance/extension/siddhi-execution-performance/component/src/main/java/org/wso2/extension/siddhi/execution/throughput/CalculatePerformanceStreamProcessorExtension.java]: I think using switch case may incur some redundant code. Is this new commit acceptable?
https://github.com/wso2/product-sp/pull/855/commits/90cdfba07f986f7d84f55a3e4133f0cefe1799cc

**Comment by Niveathika92** [comment]
@SaranganJ , @miyurud As discussed offline, I am closing this PR due to the following issues,
1. The extension cannot be built due to compile errors
2. Even though Id parameter is asked, it is not used in the file name thus it does not solve the initial issue.
3. Filename, when id param is not given, is not handled
@SaranganJ Please open a new PR after addressing the above

**Comment by Niveathika92** [comment]
Additionally, use local mkdocs server to see if the documentation is correctly deployed. Even though they are not generated since this extension is not released, it is the best practice.

**Pull request closed without merging**